examples/rust/components/http-client/src/lib.rs

```diff
@@ -0,0 +1,108 @@
+#![allow(clippy::missing_safety_doc)]
+wit_bindgen::generate!();
+
+use std::io::Read;
+
+use exports::wasi::http::incoming_handler::Guest;
+use wasi::http::types::*;
+
+#[derive(serde::Deserialize)]
+struct DogResponse {
+    message: String,
+}
+
+struct HttpServer;
+
+impl Guest for HttpServer {
+    fn handle(_request: IncomingRequest, response_out: ResponseOutparam) {
+        let req = wasi::http::outgoing_handler::OutgoingRequest::new(Fields::new());
+        req.set_scheme(Some(&Scheme::Https)).unwrap();
+        req.set_authority(Some("dog.ceo")).unwrap();
+        req.set_path_with_query(Some("/api/breeds/image/random"))
+            .unwrap();
+        let dog_picture_url = match wasi::http::outgoing_handler::handle(req, None) {
+            Ok(resp) => {
+                resp.subscribe().block();
+                let response = resp
+                    .get()
+                    .expect("HTTP request response missing")
+                    .expect("HTTP request response requested more than once")
+                    .expect("HTTP request failed");
+                if response.status() == 200 {
+                    let response_body = response
+                        .consume()
+                        .expect("failed to get incoming request body");
+                    let body = {
+                        let mut buf = vec![];
+                        let mut stream = response_body
+                            .stream()
+                            .expect("failed to get HTTP request response stream");
+                        InputStreamReader::from(&mut stream)
+                            .read_to_end(&mut buf)
+                            .expect("failed to read value from HTTP request response stream");
+                        buf
+                    };
+                    let _trailers = wasi::http::types::IncomingBody::finish(response_body);
+                    let dog_response: DogResponse = serde_json::from_slice(&body).unwrap();
+                    dog_response.message
+                } else {
+                    format!("HTTP request failed with status code {}", response.status())
+                }
+            }
+            Err(e) => {
+                format!("Got error when trying to fetch dog: {}", e)
+            }
+        };
+        let response = OutgoingResponse::new(Fields::new());
+        response.set_status_code(200).unwrap();
+        let response_body = response.body().unwrap();
+        ResponseOutparam::set(response_out, Ok(response));
+        response_body
+            .write()
```

Comment:
This should be called before `ResponseOutparam::set` - first you get a write handle and then use it after HTTP status code and headers are written